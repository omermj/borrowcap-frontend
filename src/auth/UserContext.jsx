import React from "react";

/** Context to provide currentUser thoughout the app */

const UserContext = React.createContext();

export default UserContext;
