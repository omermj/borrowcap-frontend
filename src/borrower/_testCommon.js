import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = () => {
  // Mock api call with active requests
  BorrowcapApi.getActiveRequestsByBorrowerId.mockResolvedValue([
    {
      id: 4,
      amtRequested: 10000,
      purpose: "Business",
      appOpenDate: "2021/01/01",
      interestRate: 0.05,
      term: 12,
      installmentAmt: 1000,
    },
    {
      id: 5,
      amtRequested: 20000,
      purpose: "Home",
      appOpenDate: "2021/02/01",
      interestRate: 0.07,
      term: 24,
      installmentAmt: 2000,
    },
  ]);

  BorrowcapApi.getApprovedRequestsByUserId.mockResolvedValue({
    borrower: [
      {
        id: 2,
        amtRequested: "13000",
        amtApproved: "12000",
        amtFunded: "6000",
        purpose: "Education",
        appOpenDate: "2023-01-01",
        appApprovedDate: "2023/01/05",
        fundingDeadline: "2023/02/05",
        interestRate: "0.07",
        term: 24,
        installmentAmt: "537.27",
        availableForFunding: false,
      },
    ],
  });
};

module.exports = { testUser, commonBeforeEach };
