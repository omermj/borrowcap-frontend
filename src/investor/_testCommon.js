import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getFundedLoansByUserId.mockResolvedValue({
    investor: [
      {
        id: 1,
        investorId: 3,
        amtInvested: "5000",
        amtFunded: "5000",
        fundedDate: "2023/01/01",
        interestRate: "0.05",
        term: 12,
        installmentAmt: "428.04",
        remainingBalance: "5000",
      },
    ],
  });
  await BorrowcapApi.getAvailableInvestments.mockResolvedValue([
    {
      id: 7,
      amtApproved: "8000",
      amtFunded: "0",
      purpose: "Medical",
      approvedDate: "2023/01/01",
      fundingDeadline: "2023/02/1",
      interestRate: "0.07",
      term: 24,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
