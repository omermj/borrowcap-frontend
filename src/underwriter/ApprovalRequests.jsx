import { useState, useEffect } from "react";
import TableComponent from "../common/TableComponent";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import BorrowcapApi from "../api/api";

/** Table showing all Active Loan Requests for logged in user */

const ApprovalRequests = () => {
  const [loanRequests, setLoanRequests] = useState([]);

  // Define table headers with labels and formatters
  const headers = {
    id: { label: "App ID", formatter: "none" },
    borrowerId: { label: "Borrower ID", formatter: "none" },
    amtRequested: {
      label: "Requested Amount",
      formatter: formatCurrency,
    },
    purpose: { label: "Purpose", formatter: "none" },
    appOpenDate: { label: "Application Date", formatter: formatDate },
    interestRate: { label: "Interest Rate", formatter: formatPercent },
    term: { label: "Term", formatter: "none" },
    installmentAmt: { label: "Installment", formatter: formatCurrency },
    reviewBtn: {
      label: "Review",
      formatter: "button",
      link: "/underwriter/reviewrequest",
      onClick: null,
    },
  };

  // get active loan requests on initial render
  useEffect(() => {
    async function fetchLoanRequests() {
      const requests = await BorrowcapApi.getLoanRequests();
      setLoanRequests(requests);
    }
    fetchLoanRequests();
  }, []);

  if (!loanRequests.length) return <div></div>;

  return (
    <div>
      <p>Approval Requests</p>
      <TableComponent headers={headers} tableData={loanRequests} />
    </div>
  );
};

export default ApprovalRequests;
