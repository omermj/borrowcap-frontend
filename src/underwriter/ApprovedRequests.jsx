import { useState, useEffect } from "react";
import TableComponent from "../common/TableComponent";
import TableHeader from "../common/Table/TableHeader";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import BorrowcapApi from "../api/api";

/** Table showing all Active Loan Requests for logged in user */

const ApprovedRequests = () => {
  const [data, setData] = useState([]);

  // Define table headers with labels and formatters
  const headers = {
    id: { label: "App ID", formatter: "none" },
    borrowerId: { label: "Borrower ID", formatter: "none" },
    amtRequested: {
      label: "Requested Amount",
      formatter: formatCurrency,
    },
    amtApproved: {
      label: "Approved Amount",
      formatter: formatCurrency,
    },
    purpose: { label: "Purpose", formatter: "none" },
    appApprovedDate: { label: "Approval Date", formatter: formatDate },
    interestRate: { label: "Interest Rate", formatter: formatPercent },
    term: { label: "Term", formatter: "none" },
    installmentAmt: { label: "Installment", formatter: formatCurrency },
  };

  // get active loan requests on initial render
  useEffect(() => {
    async function fetchApprovedRequests() {
      const requests = await BorrowcapApi.getApprovedRequests();
      setData(requests);
    }
    fetchApprovedRequests();
  }, []);

  return (
    <div className="border mb-4">
      <TableHeader text={"Approved Requests"} />
      {!data.length ? (
        <div className=" mb-2 fst-italic fw-light">
          {" "}
          <small>No approved requests.</small>
        </div>
      ) : (
        <TableComponent headers={headers} tableData={data} />
      )}
    </div>
  );
};

export default ApprovedRequests;
