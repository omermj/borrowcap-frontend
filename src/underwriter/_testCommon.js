import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getLoanRequests.mockResolvedValue([
    {
      id: 2,
      borrowerId: 2,
      amtRequested: "10000",
      purpose: "Education",
      purposeId: 3,
      appOpenDate: "2023/01/01",
      interestRate: "0.07",
      term: 36,
      installmentAmt: "319.86",
    },
  ]);

  await BorrowcapApi.getApprovedRequests.mockResolvedValue([
    {
      id: 1,
      borrowerId: 2,
      amtRequested: "10000",
      amtApproved: "9000",
      amtFunded: "5000",
      purpose: "Car",
      purposeId: 2,
      appOpenDate: "2023/01/01",
      appApprovedDate: "2023/01/05",
      fundingDeadline: "2023/02/05",
      interestRate: "0.05",
      term: 12,
      installmentAmt: "428.04",
      availableForFunding: false,
      isFunded: true,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
