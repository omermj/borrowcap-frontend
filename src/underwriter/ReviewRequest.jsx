import { Container, Row, Col } from "react-bootstrap";
import { useState, useEffect, useContext } from "react";
import { useParams, useNavigate } from "react-router-dom";
import BorrowcapApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import { Formik } from "formik";
import { Form, Button } from "react-bootstrap";
import UserContext from "../auth/UserContext";

/** Displays a single Available Investment */

const ReviewRequest = ({ terms }) => {
  const { currentUser } = useContext(UserContext);
  const { id } = useParams();
  const [data, setData] = useState(null);
  const navigate = useNavigate();

  useEffect(() => {
    const getReviewRequest = async () => {
      const result = await BorrowcapApi.getActiveRequest(id);
      if (result) {
        setData(result);
      }
    };
    getReviewRequest();
  }, [id]);

  if (!data) return <LoadingSpinner />;

  const handleReject = async () => {
    const res = await BorrowcapApi.rejectRequest(id);
    if (!!res) navigate("/underwriter");
  };

  return (
    <div>
      {console.log(data)}
      <h3>Loan Request Details</h3>
      <Container>
        <Row>
          <Col>
            <div>Amount Requested: {formatCurrency(data.amtRequested)}</div>
            <div>Loan Purpose: {data.purpose} </div>
            <div>Application Open Date: {formatDate(data.appOpenDate)}</div>
            <div>Term: {data.term} months </div>
            <div>Interest Rate: {formatPercent(data.interestRate)} </div>
            <div>Installment Amount: {formatCurrency(data.installmentAmt)}</div>
          </Col>
          <Col>
            <div>Username: {data.username}</div>
            <div>First Name: {data.firstName}</div>
            <div>Last Name: {data.lastName}</div>
            <div>Email: {data.email}</div>
            <div>Account Balance: {data.accountBalance}</div>
            <div>
              Borrower&apos;s Income: {formatCurrency(data.annualIncome)}
            </div>
            <div>
              Other Monthly Debt: {formatCurrency(data.otherMonthlyDebt)}
            </div>
          </Col>
        </Row>
        <Row>
          <Col>
            <Formik
              initialValues={{
                amtApproved: "",
                interestRate: "",
                term: "",
              }}
              onSubmit={async (values, { setSubmitting, setErrors }) => {
                const res = await BorrowcapApi.approveRequest(id, values);
                if (!!res) navigate("/underwriter");
              }}
            >
              {({
                values,
                errors,
                handleChange,
                handleSubmit,
                isSubmitting,
              }) => (
                <Form onSubmit={handleSubmit}>
                  <Form.Group className="mb-3" controlId="approvalFormAmount">
                    <Form.Label>Approved Amount</Form.Label>
                    <Form.Control
                      type="number"
                      name="amtApproved"
                      min={0}
                      value={values.amtApproved}
                      onChange={handleChange}
                      required
                    />
                  </Form.Group>
                  <Form.Group className="mb-3" controlId="approvalFormInterest">
                    <Form.Label>Interest Rate</Form.Label>
                    <Form.Control
                      type="number"
                      name="interestRate"
                      min={0}
                      max={1}
                      step={0.001}
                      value={values.interestRate}
                      onChange={handleChange}
                      required
                    />
                  </Form.Group>
                  <Form.Group className="mb-3" controlId="approvalFormTerm">
                    <Form.Label>Term</Form.Label>
                    <Form.Select
                      aria-label="Term Select"
                      name="term"
                      onChange={handleChange}
                    >
                      <option>Select loan term</option>
                      {terms.map((term, idx) => (
                        <option key={idx} value={term.months}>
                          {term.months} months
                        </option>
                      ))}
                    </Form.Select>
                  </Form.Group>
                  <Button
                    variant="success"
                    type="submit"
                    disabled={isSubmitting}
                    className="mt-4"
                  >
                    Approve
                  </Button>
                  <Button
                    variant="danger"
                    disabled={isSubmitting}
                    className="mt-4"
                    onClick={handleReject}
                  >
                    Reject
                  </Button>
                </Form>
              )}
            </Formik>
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default ReviewRequest;
