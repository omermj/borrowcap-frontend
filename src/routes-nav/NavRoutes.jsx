import { Route, Routes } from "react-router-dom";
import { useSelector } from "react-redux";
import BorrowerMain from "../borrower/BorrowerMain";
import Homepage from "../homepage/Homepage";
import InvestorMain from "../investor/InvestorMain";
import UnderwriterMain from "../underwriter/UnderwriterMain";
import LoginForm from "../auth/LoginForm";
import SignupForm from "../auth/SignupForm";
import ProtectedRoute from "./ProtectedRoute";
import LoanApplicationForm from "../borrower/LoanApplication";
import AvailableInvestment from "../investor/AvailableInvestment";
import ReviewRequest from "../underwriter/ReviewRequest";
import ApprovedLoan from "../borrower/ApprovedRequest";
import ActiveRequests from "../borrower/ActiveRequests";
import ApprovedLoans from "../borrower/ApprovedRequests";
import FundedLoans from "../borrower/FundedLoans";
import AvailableInvestments from "../investor/AvailableInvestments";
import PledgedInvestments from "../investor/PledgedInvestments";
import ActiveInvestments from "../investor/ActiveInvestments";
import ApprovalRequests from "../underwriter/ApprovalRequests";
import ApprovedRequests from "../underwriter/ApprovedRequests";
import Wallet from "../common/Wallet";
import ProfileUpdateForm from "../common/ProfileUpdateForm";
import ChangePasswordForm from "../common/ChangePassword";
import NotFound from "../common/NotFound";

/** Site-wide routes */

function NavRoutes({ signup, roles }) {
  const currentUser = useSelector((state) => state.userState.user);

  const homepage = () => {
    if (currentUser) {
      if (currentUser.roles.includes("admin")) {
        return <UnderwriterMain />;
      } else if (currentUser.roles.includes("borrower")) {
        return <BorrowerMain />;
      } else if (currentUser.roles.includes("investor")) {
        return <InvestorMain />;
      }
    }
    return <Homepage />;
  };

  return (
    <div>
      <Routes>
        <Route path="/" element={homepage()} />

        {/* --------------- */}
        {/* Borrower Routes */}
        {/* --------------- */}
        <Route element={<ProtectedRoute allowedRoles={["borrower"]} />}>
          <Route path="/borrower" element={<BorrowerMain />} />
          <Route path="/borrower/apply" element={<LoanApplicationForm />} />
          <Route
            path="/borrower/approvedrequests/:id"
            element={<ApprovedLoan />}
          />
          <Route path="/borrower/activerequests" element={<ActiveRequests />} />
          <Route path="/borrower/approvedloans" element={<ApprovedLoans />} />
          <Route path="/borrower/fundedloans" element={<FundedLoans />} />
          <Route path="/borrower/wallet" element={<Wallet />} />
          <Route path="/borrower/profile" element={<ProfileUpdateForm />} />
        </Route>

        {/* --------------- */}
        {/* Investor Routes */}
        {/* --------------- */}
        <Route element={<ProtectedRoute allowedRoles={["investor"]} />}>
          <Route path="/investor" element={<InvestorMain />} />
          <Route
            path="/investor/availableinvestment/:id"
            element={<AvailableInvestment />}
          />
          <Route
            path="/investor/availableinvestments"
            element={<AvailableInvestments />}
          />
          <Route path="/investor/pledges" element={<PledgedInvestments />} />
          <Route
            path="/investor/activeinvestments"
            element={<ActiveInvestments />}
          />
          <Route path="/investor/wallet" element={<Wallet />} />
          <Route path="/investor/profile" element={<ProfileUpdateForm />} />
        </Route>

        {/* ------------------------ */}
        {/* Underwriter/Admin Routes */}
        {/* ------------------------ */}
        <Route element={<ProtectedRoute allowedRoles={["admin"]} />}>
          <Route path="/underwriter" element={<UnderwriterMain />} />
          <Route
            path="/underwriter/reviewrequest/:id"
            element={<ReviewRequest />}
          />
          <Route path="/underwriter/requests" element={<ApprovalRequests />} />
          <Route path="/underwriter/approved" element={<ApprovedRequests />} />
        </Route>

        {/* ------------------------ */}
        {/* Common User Routes */}
        {/* ------------------------ */}
        <Route
          element={
            <ProtectedRoute allowedRoles={["admin", "borrower", "investor"]} />
          }
        >
          <Route path="/user/profile" element={<ProfileUpdateForm />} />
          <Route path="/user/changepassword" element={<ChangePasswordForm />} />
        </Route>

        {/* ----------- */}
        {/* Auth Routes */}
        {/* ----------- */}
        <Route path="/login" element={<LoginForm />} />
        <Route
          path="/signup"
          element={<SignupForm signup={signup} roles={roles} />}
        />

        {/* 404 Not Found */}
        <Route path="*" element={<NotFound />} />
      </Routes>
    </div>
  );
}

export default NavRoutes;
