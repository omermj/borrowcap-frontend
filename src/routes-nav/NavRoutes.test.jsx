import { render } from "@testing-library/react";
import NavRoutes from "./NavRoutes";
import { BrowserRouter } from "react-router-dom";
import UserContext from "../auth/UserContext";
import { commonBeforeEach } from "../underwriter/_testCommon";

// Mock api calls
vi.mock("../api/api");
beforeEach(commonBeforeEach);

// smoke test
it("renders without crashing", () => {
  render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: { username: "test", roles: ["admin"] } }}
      >
        <NavRoutes roles={["admin"]} purposes={"Home"} terms={"24"} />
      </UserContext.Provider>
    </BrowserRouter>
  );
});

// snapshot test
it("matches snapshot", () => {
  const { asFragment } = render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: { username: "test", roles: ["admin"] } }}
      >
        <NavRoutes roles={["admin"]} purposes={"Home"} terms={"24"} />
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
