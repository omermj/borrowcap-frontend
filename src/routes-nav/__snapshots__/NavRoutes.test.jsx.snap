// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        class="border mb-4"
      >
        <div
          class="py-2"
        >
          <span
            class="align-middle h5"
          >
            Approval Requests
          </span>
        </div>
        <div
          class=" mb-2 fst-italic fw-light"
        >
           
          <small>
            No approval requests.
          </small>
        </div>
      </div>
      <div
        class="border mb-4"
      >
        <div
          class="py-2"
        >
          <span
            class="align-middle h5"
          >
            Approved Requests
          </span>
        </div>
        <div
          class=" mb-2 fst-italic fw-light"
        >
           
          <small>
            No approved requests.
          </small>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
